# Openclassrooms PJ3 : IMDB dataset :  data cleaning notebook 

In [65]:
%matplotlib inline

import os
import zipfile
import urllib

import matplotlib.pyplot as plt

import numpy as np

import qgrid

from pandas.plotting import scatter_matrix

DOWNLOAD_ROOT = "https://s3-eu-west-1.amazonaws.com/static.oc-static.com/prod/courses/files/Parcours_data_scientist/Projet+-+Moteur+de+recommandation+de+films/"
DATA_PATH = os.path.join("datasets", "imdb")

DATA_PATH_FILE = os.path.join(DATA_PATH, "movie_metadata.csv")
DATA_URL = DOWNLOAD_ROOT + "imdb-5000-movie-dataset.zip"

DATA_PATH_FILE_OUTPUT = os.path.join(DATA_PATH, "movie_metadata_transformed.csv")

DOWNLOAD_DATA = False  # A la première exécution du notebook, ou pour rafraîchir les données, mettre cette variable à True

plt.rcParams["figure.figsize"] = [16,9] # Taille par défaut des figures de matplotlib

import seaborn as sns
sns.set()


pd.set_option('display.max_columns', None)


# Téléchargement et décompression des données

In [2]:
PROXY_DEF = 'BNP'
#PROXY_DEF = None

def fetch_dataset(data_url=DATA_URL, data_path=DATA_PATH):
    if not os.path.isdir(data_path):
        os.makedirs(data_path)

    archive_path = os.path.join(data_path, "imdb-5000-movie-dataset.zip")
    
    if (PROXY_DEF == 'BNP'):
        #create the object, assign it to a variable
        proxy = urllib.request.ProxyHandler({'https': 'https://login:password@ncproxy:8080'})
        # construct a new opener using your proxy settings
        opener = urllib.request.build_opener(proxy)
        # install the openen on the module-level
        urllib.request.install_opener(opener)    
    
    urllib.request.urlretrieve(data_url, archive_path)
    data_archive = zipfile.ZipFile(archive_path)
    data_archive.extractall(path=data_path)
    data_archive.close()

In [3]:
if (DOWNLOAD_DATA == True):
    fetch_dataset()

# Import du fichier CSV

## Inspection de quelques lignes du fichier pour avoir un aperçu visuel du texte brut :

In [4]:
def read_raw_file(nblines, data_path = DATA_PATH):
    csv_path = DATA_PATH_FILE
    
    fp = open(csv_path)
    
    line = ""
    
    for cnt_lines in range(nblines+1):
        line = fp.readline()
        
    print(">>>>>> Line %d" % (cnt_lines))
    print(line)
    
    

In [5]:
read_raw_file(0)
read_raw_file(1)
read_raw_file(2)

>>>>>> Line 0
color,director_name,num_critic_for_reviews,duration,director_facebook_likes,actor_3_facebook_likes,actor_2_name,actor_1_facebook_likes,gross,genres,actor_1_name,movie_title,num_voted_users,cast_total_facebook_likes,actor_3_name,facenumber_in_poster,plot_keywords,movie_imdb_link,num_user_for_reviews,language,country,content_rating,budget,title_year,actor_2_facebook_likes,imdb_score,aspect_ratio,movie_facebook_likes

>>>>>> Line 1
Color,James Cameron,723,178,0,855,Joel David Moore,1000,760505847,Action|Adventure|Fantasy|Sci-Fi,CCH Pounder,Avatar ,886204,4834,Wes Studi,0,avatar|future|marine|native|paraplegic,http://www.imdb.com/title/tt0499549/?ref_=fn_tt_tt_1,3054,English,USA,PG-13,237000000,2009,936,7.9,1.78,33000

>>>>>> Line 2
Color,Gore Verbinski,302,169,563,1000,Orlando Bloom,40000,309404152,Action|Adventure|Fantasy,Johnny Depp,Pirates of the Caribbean: At World's End ,471220,48350,Jack Davenport,0,goddess|marriage ceremony|marriage proposal|pirate|singapore,http://ww

## Chargement des données

In [6]:
import pandas as pd

def load_data(data_path=DATA_PATH):
    csv_path = DATA_PATH_FILE
    return pd.read_csv(csv_path, sep=',', header=0, encoding='utf-8')


In [7]:
df = load_data()

###  On vérifie que le nombre de lignes intégrées dans le Dataframe correspond au nombre de lignes du fichier

In [8]:
num_lines = sum(1 for line in open(DATA_PATH_FILE, encoding='utf-8'))
message = (
f"Nombre de lignes dans le fichier (en comptant l'entête): {num_lines}\n"
f"Nombre d'instances dans le dataframe: {df.shape[0]}"
)
print(message)

Nombre de lignes dans le fichier (en comptant l'entête): 5044
Nombre d'instances dans le dataframe: 5043


### Puis on affiche quelques instances de données :

In [9]:
df.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                  NaN      NaN      NaN             NaN          NaN   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

In [10]:
def qgrid_show(df):
    display(qgrid.show_grid(df, grid_options={'forceFitColumns': False, 'defaultColumnWidth': 170}))

In [11]:
display(qgrid.show_grid(df, grid_options={'forceFitColumns': False, 'defaultColumnWidth': 170}))

QgridWidget(grid_options={'fullWidthRows': True, 'syncColumnCellResize': True, 'forceFitColumns': False, 'defa…

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5043 entries, 0 to 5042
Data columns (total 28 columns):
color                        5024 non-null object
director_name                4939 non-null object
num_critic_for_reviews       4993 non-null float64
duration                     5028 non-null float64
director_facebook_likes      4939 non-null float64
actor_3_facebook_likes       5020 non-null float64
actor_2_name                 5030 non-null object
actor_1_facebook_likes       5036 non-null float64
gross                        4159 non-null float64
genres                       5043 non-null object
actor_1_name                 5036 non-null object
movie_title                  5043 non-null object
num_voted_users              5043 non-null int64
cast_total_facebook_likes    5043 non-null int64
actor_3_name                 5020 non-null object
facenumber_in_poster         5030 non-null float64
plot_keywords                4890 non-null object
movie_imdb_link              5043 non-

In [13]:
df.describe()

num_critic_for_reviews     duration  director_facebook_likes  \
count             4993.000000  5028.000000              4939.000000   
mean               140.194272   107.201074               686.509212   
std                121.601675    25.197441              2813.328607   
min                  1.000000     7.000000                 0.000000   
25%                 50.000000    93.000000                 7.000000   
50%                110.000000   103.000000                49.000000   
75%                195.000000   118.000000               194.500000   
max                813.000000   511.000000             23000.000000   

       actor_3_facebook_likes  actor_1_facebook_likes         gross  \
count             5020.000000             5036.000000  4.159000e+03   
mean               645.009761             6560.047061  4.846841e+07   
std               1665.041728            15020.759120  6.845299e+07   
min                  0.000000                0.000000  1.620000e+02   
25%                133.000000              614.000000  5.340988e+06   
50%                371.500000              988.000000  2.551750e+07   
75%                636.000000            11000.000000  6.230944e+07   
max              23000.000000           640000.000000  7.605058e+08   

       num_voted_users  cast_total_facebook_likes  facenumber_in_poster  \
count     5.043000e+03                5043.000000           5030.000000   
mean      8.366816e+04                9699.063851              1.371173   
std       1.384853e+05               18163.799124              2.013576   
min       5.000000e+00                   0.000000              0.000000   
25%       8.593500e+03                1411.000000              0.000000   
50%       3.435900e+04                3090.000000              1.000000   
75%       9.630900e+04               13756.500000              2.000000   
max       1.689764e+06              656730.000000             43.000000   

       num_user_for_reviews        budget   title_year  \
count           5022.000000  4.551000e+03  4935.000000   
mean             272.770808  3.975262e+07  2002.470517   
std              377.982886  2.061149e+08    12.474599   
min                1.000000  2.180000e+02  1916.000000   
25%               65.000000  6.000000e+06  1999.000000   
50%              156.000000  2.000000e+07  2005.000000   
75%              326.000000  4.500000e+07  2011.000000   
max             5060.000000  1.221550e+10  2016.000000   

       actor_2_facebook_likes   imdb_score  aspect_ratio  movie_facebook_likes  
count             5030.000000  5043.000000   4714.000000           5043.000000  
mean              1651.754473     6.442138      2.220403           7525.964505  
std               4042.438863     1.125116      1.385113          19320.445110  
min                  0.000000     1.600000      1.180000              0.000000  
25%                281.000000     5.800000      1.850000              0.000000  
50%                595.000000     6.600000      2.350000            166.000000  
75%                918.000000     7.200000      2.350000           3000.000000  
max             137000.000000     9.500000     16.000000         349000.000000

### Vérification s'il y a des doublons

In [14]:
df[df.duplicated()]

color         director_name  num_critic_for_reviews  \
137              Color           David Yates                   248.0   
187              Color           Bill Condon                   322.0   
204              Color          Hideaki Anno                     1.0   
303              Color            Joe Wright                   256.0   
389              Color            Josh Trank                   369.0   
395              Color             Rob Cohen                   187.0   
590              Color          Brett Ratner                   245.0   
656              Color        Paul Verhoeven                   196.0   
794              Color           Joss Whedon                   703.0   
1220             Color   Angelina Jolie Pitt                   322.0   
1305             Color         Paul McGuigan                   159.0   
1449             Color         Albert Hughes                   208.0   
2169             Color         Paul McGuigan                    98.0   
2292             Color              Frank Oz                   168.0   
2472             Color            Jon Cassar                    45.0   
2493   Black and White           Yimou Zhang                   283.0   
2533             Color           Neil Burger                   236.0   
2562             Color             Jon Lucas                    81.0   
2568             Color         Vic Armstrong                   169.0   
2619             Color        John Carpenter                   318.0   
2771             Color          Ole Bornedal                   264.0   
2777             Color        Stephen Frears                    51.0   
2798             Color            Shawn Levy                    69.0   
2971             Color      John Lee Hancock                   106.0   
3117             Color           Guy Ritchie                   151.0   
3345             Color          Herbert Ross                    60.0   
3452             Color           Paul Haggis                   287.0   
3480             Color  Michael Winterbottom                    71.0   
3729             Color      Tim Blake Nelson                    92.0   
3900             Color                   NaN                     9.0   
3915             Color            Wes Craven                   160.0   
4182             Color            Rob Zombie                   220.0   
4226             Color      William Friedkin                   138.0   
4282             Color       Kenneth Branagh                    85.0   
4313             Color       Bruce McCulloch                    52.0   
4408             Color           Yimou Zhang                   101.0   
4565             Color        Peter Cattaneo                   122.0   
4573             Color            Mel Brooks                    48.0   
4631             Color           Danny Boyle                   393.0   
4769             Color           Tamra Davis                   111.0   
4882             Color            Dan Curtis                     NaN   
4927             Color           Jason Stone                    48.0   
4942             Color         Paul Schrader                   130.0   
4950             Color         David Hewlett                     8.0   
4951   Black and White      George A. Romero                   284.0   

      duration  director_facebook_likes  actor_3_facebook_likes  \
137      110.0                    282.0                   103.0   
187      115.0                    386.0                 12000.0   
204      120.0                     28.0                    12.0   
303      111.0                    456.0                   394.0   
389      100.0                    128.0                    78.0   
395      106.0                    357.0                  4000.0   
590      101.0                    420.0                   467.0   
656      113.0                    719.0                   217.0   
794      173.0                      0.0                 19000.0   
1220     137.0                  11

### Suppression des doublons

In [15]:
df.drop_duplicates(inplace=True)

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4998 entries, 0 to 5042
Data columns (total 28 columns):
color                        4979 non-null object
director_name                4895 non-null object
num_critic_for_reviews       4949 non-null float64
duration                     4983 non-null float64
director_facebook_likes      4895 non-null float64
actor_3_facebook_likes       4975 non-null float64
actor_2_name                 4985 non-null object
actor_1_facebook_likes       4991 non-null float64
gross                        4124 non-null float64
genres                       4998 non-null object
actor_1_name                 4991 non-null object
movie_title                  4998 non-null object
num_voted_users              4998 non-null int64
cast_total_facebook_likes    4998 non-null int64
actor_3_name                 4975 non-null object
facenumber_in_poster         4985 non-null float64
plot_keywords                4846 non-null object
movie_imdb_link              4998 non-

Imputation des variables manquantes :
    https://scikit-learn.org/stable/modules/generated/sklearn.impute.SimpleImputer.html
https://scikit-learn.org/stable/modules/impute.html  => à lire en premier

ACP et 1 hot encoding :  y a-t-il une autre possibilité ?

Valeurs de variables très peu représentées :   => à voir dans un second temps
https://machinelearningmastery.com/tactics-to-combat-imbalanced-classes-in-your-machine-learning-dataset/


## Affichage des champs renseignés (non NA) avec leur pourcentage de complétude
L'objectif est de voir quelles sont les features qui seront les plus fiables en terme de qualité de donnée, et quelles sont celles pour lesquelles on devra faire des choix

In [17]:
(df.count()/df.shape[0]).sort_values(axis=0, ascending=False)

movie_facebook_likes         1.000000
movie_imdb_link              1.000000
genres                       1.000000
movie_title                  1.000000
num_voted_users              1.000000
cast_total_facebook_likes    1.000000
imdb_score                   1.000000
country                      0.999000
actor_1_name                 0.998599
actor_1_facebook_likes       0.998599
language                     0.997599
actor_2_facebook_likes       0.997399
actor_2_name                 0.997399
facenumber_in_poster         0.997399
duration                     0.996999
color                        0.996198
num_user_for_reviews         0.995798
actor_3_name                 0.995398
actor_3_facebook_likes       0.995398
num_critic_for_reviews       0.990196
director_facebook_likes      0.979392
director_name                0.979392
title_year                   0.978591
plot_keywords                0.969588
content_rating               0.939776
aspect_ratio                 0.934574
budget      

## Identification des typologies de features à traiter 

In [133]:
numerical_features = ['movie_facebook_likes', 'num_voted_users', 'cast_total_facebook_likes', 'imdb_score' , 'actor_1_facebook_likes', 'actor_2_facebook_likes', 'facenumber_in_poster', 'duration', 'num_user_for_reviews', 'actor_3_facebook_likes', 'num_critic_for_reviews', 'director_facebook_likes', 'budget', 'gross','title_year']

# à 1 hot encoder, et à splitter avant si nécessaire  ('genres' et 'plot_keywords' doivent être splittées)
categorical_features = ['country', 'director_name', 'genres', 'plot_keywords', 'color', 'content_rating']

# à transformer en bag of words
categorical_features_tobow = ['movie_title']  

# à fusioner en 1 seule variable
categorical_features_tomerge = ['actor_1_name', 'actor_2_name', 'actor_3_name']  

# features qui ne seront pas conservées :
features_notkept = ['aspect_ratio', 'movie_imdb_link']





## Affichage des features qui seront splittées avant le 1hot encode :

In [150]:
df[['genres', 'plot_keywords']]

genres  \
0     Action|Adventure|Fantasy|Sci-Fi   
1            Action|Adventure|Fantasy   
2           Action|Adventure|Thriller   
3                     Action|Thriller   
4                         Documentary   
...                               ...   
5038                     Comedy|Drama   
5039     Crime|Drama|Mystery|Thriller   
5040            Drama|Horror|Thriller   
5041             Comedy|Drama|Romance   
5042                      Documentary   

                                          plot_keywords  
0                avatar|future|marine|native|paraplegic  
1     goddess|marriage ceremony|marriage proposal|pi...  
2                   bomb|espionage|sequel|spy|terrorist  
3     deception|imprisonment|lawlessness|police offi...  
4                                                   NaN  
...                                                 ...  
5038             fraud|postal worker|prison|theft|trial  
5039       cult|fbi|hideout|prison escape|serial killer  
5040                                                NaN  
5041                                                NaN  
5042  actress name in title|crush|date|four word tit...  

[4998 rows x 2 columns]

# Imputation des données manquantes

In [77]:
# KNN imputer pas encore supporté par la version de sklearn que j'utilise :

#from sklearn.impute import KNNImputer

#imputer = KNNImputer(n_neighbors=2, weights="uniform")  
#imputer.fit_transform(df[numerical_features])

In [78]:
numerical_features_columns = df[numerical_features].columns

In [79]:
numerical_features_index = df[numerical_features].index

In [80]:
numerical_features_columns.shape

(15,)

## Imputation des données numériques par régression linéaire

In [81]:
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer
imp = IterativeImputer(max_iter=10, random_state=0)

transformed_data = imp.fit_transform(df[numerical_features])  



In [82]:
df_numerical_features_imputed = pd.DataFrame(data=transformed_data, columns=numerical_features_columns, index=numerical_features_index)

### Visualisation de quelques résultats par comparaison avant/après :

In [84]:
qgrid_show(df[numerical_features])

QgridWidget(grid_options={'fullWidthRows': True, 'syncColumnCellResize': True, 'forceFitColumns': False, 'defa…

In [86]:
qgrid_show(df_numerical_features_imputed)

QgridWidget(grid_options={'fullWidthRows': True, 'syncColumnCellResize': True, 'forceFitColumns': False, 'defa…

### Constat que toutes les valeurs sont maintenant renseignées :

In [161]:
(df_numerical_features_imputed.count()/df_numerical_features_imputed.shape[0]).sort_values(axis=0, ascending=False)

title_year                   1.0
gross                        1.0
budget                       1.0
director_facebook_likes      1.0
num_critic_for_reviews       1.0
actor_3_facebook_likes       1.0
num_user_for_reviews         1.0
duration                     1.0
facenumber_in_poster         1.0
actor_2_facebook_likes       1.0
actor_1_facebook_likes       1.0
imdb_score                   1.0
cast_total_facebook_likes    1.0
num_voted_users              1.0
movie_facebook_likes         1.0
dtype: float64

## Transformation des features de catégorie
### Voir le §  Identification des typologies de features à traiter  ci-dessus pour une explication des différents cas de figure

In [149]:
'''
Cette fonction fait un 1 hot encoding des features qui sont des catégories
Elle fonctionne pour les 2 cas de figure suivant :
- Les valeurs possibles de la colonne sont une chaîne de caractère (ex : cat1)
- Les valeurs possibles de la colonne sont des chaînes de caractère avec des séparateurs (ex:  cat1|cat2|cat3)
'''

def add_categorical_features_1hot(df, df_target, categorical_features_totransform):
    for feature_totransform in categorical_features_totransform:
        print(f'Adding 1hot Feature : {feature_totransform}')
        df_transformed = df[feature_totransform].str.get_dummies().add_prefix(feature_totransform +'_')
        df_target = pd.concat([df_target, df_transformed], axis=1)
        
    return(df_target)

'''
Cette fonction commence par merger les valeurs de toutes les colonnes comprises dans 
categorical_features_tomerge_andtransform  dans une colonne temporaire
Puis elle fait un 1 hot encode du résultat en appelant la fonction add_categorical_features_1hot

df :  dataframe source
df_target : dataframe cible où seront concaténées les nouvelles features créées
categorical_features_tomerge_andtransform : la liste des catégories à merger et à 1 hot encode,
             par exemple: ['actor_1_name', 'actor_2_name', 'actor_3_name']
merged_feature_name : le nom de la feature qui sera mergée
    exemple si le nom est 'actors_names'
    On pourra avoir les colonnes suivantes de créées:  'actors_names_Le nom du 1er acteur', 'actors_names_Le nom du 2eme acteur'
          
'''
def add_categorical_features_merge_and_1hot(df, df_target, categorical_features_tomerge_andtransform, merged_feature_name):
    cnt = 0
    for feature_totransform in categorical_features_tomerge_andtransform:                            
        if (cnt == 0):
            df[merged_feature_name] = df[feature_totransform]
        
        else:
            df[merged_feature_name] = df[merged_feature_name] + '|' + df[feature_totransform]
            
        cnt += 1
        
    return(add_categorical_features_1hot(df, df_target, [merged_feature_name]))
            

In [110]:
df_imputed = add_categorical_features_1hot(df, df_numerical_features_imputed, categorical_features)
df_imputed = add_categorical_features_merge_and_1hot(df, df_imputed, categorical_features_tomerge, 'actors_names' )

Feature : country
Feature : director_name
Feature : genres
Feature : plot_keywords
Feature : color
Feature : content_rating


## Comparaison avant/après de quelques valeurs 1hot encoded :

In [159]:
df[['actor_1_name', 'actor_2_name', 'actor_3_name', 'authors_names', 'country', 'genres']].head(10)

actor_1_name       actor_2_name          actor_3_name  \
0      CCH Pounder   Joel David Moore             Wes Studi   
1      Johnny Depp      Orlando Bloom        Jack Davenport   
2  Christoph Waltz       Rory Kinnear      Stephanie Sigman   
3        Tom Hardy     Christian Bale  Joseph Gordon-Levitt   
4      Doug Walker         Rob Walker                   NaN   
5     Daryl Sabara    Samantha Morton          Polly Walker   
6     J.K. Simmons       James Franco         Kirsten Dunst   
7     Brad Garrett       Donna Murphy           M.C. Gainey   
8  Chris Hemsworth  Robert Downey Jr.    Scarlett Johansson   
9     Alan Rickman   Daniel Radcliffe          Rupert Grint   

                                       authors_names country  \
0             CCH Pounder|Joel David Moore|Wes Studi     USA   
1           Johnny Depp|Orlando Bloom|Jack Davenport     USA   
2      Christoph Waltz|Rory Kinnear|Stephanie Sigman      UK   
3      Tom Hardy|Christian Bale|Joseph Gordon-Levitt     USA   
4                                                NaN     NaN   
5          Daryl Sabara|Samantha Morton|Polly Walker     USA   
6            J.K. Simmons|James Franco|Kirsten Dunst     USA   
7              Brad Garrett|Donna Murphy|M.C. Gainey     USA   
8  Chris Hemsworth|Robert Downey Jr.|Scarlett Joh...     USA   
9         Alan Rickman|Daniel Radcliffe|Rupert Grint      UK   

                                              genres  
0                    Action|Adventure|Fantasy|Sci-Fi  
1                           Action|Adventure|Fantasy  
2                          Action|Adventure|Thriller  
3                                    Action|Thriller  
4                                        Documentary  
5                            Action|Adventure|Sci-Fi  
6                           Action|Adventure|Romance  
7  Adventure|Animation|Comedy|Family|Fantasy|Musi...  
8                            Action|Adventure|Sci-Fi  
9                   Adventure|Family|Fantasy|Mystery

In [160]:
df_imputed[['authors_names_Johnny Depp', 'authors_names_Orlando Bloom', 'authors_names_Jack Davenport', 'authors_names_Joel David Moore', 'country_USA', 'country_UK', 'genres_Action', 'genres_Adventure']].head(10)

authors_names_Johnny Depp  authors_names_Orlando Bloom  \
0                          0                            0   
1                          1                            1   
2                          0                            0   
3                          0                            0   
4                          0                            0   
5                          0                            0   
6                          0                            0   
7                          0                            0   
8                          0                            0   
9                          0                            0   

   authors_names_Jack Davenport  authors_names_Joel David Moore  country_USA  \
0                             0                               1            1   
1                             1                               0            1   
2                             0                               0            0   
3                             0                               0            1   
4                             0                               0            0   
5                             0                               0            1   
6                             0                               0            1   
7                             0                               0            1   
8                             0                               0            1   
9                             0                               0            0   

   country_UK  genres_Action  genres_Adventure  
0           0              1                 1  
1           0              1                 1  
2           1              1                 1  
3           0              1                 0  
4           0              0                 0  
5           0              1                 1  
6           0              1                 1  
7           0              0                 1  
8           0              1                 1  
9           1              0                 1

In [102]:
df_imputed[df_imputed['actor_1_name_Joel David Moore'] == 1][['actor_1_name_Joel David Moore']].head(30)

actor_1_name_Joel David Moore
4426                              1

In [ ]:
df.hist(bins=50, figsize=(20,15))

In [ ]:
scatter_matrix(df[numerical_features], figsize=(30,30))
plt.suptitle('Diagramme de dispersion des données numériques')

In [ ]:
corr_matrix = df.corr()

In [ ]:
corr_matrix[numerical_features].loc[numerical_features]

In [ ]:

plt.title('Corrélation entre les valeurs numériques')
sns.heatmap(corr_matrix[numerical_features].loc[numerical_features], 
        xticklabels=corr_matrix[numerical_features].loc[numerical_features].columns,
        yticklabels=corr_matrix[numerical_features].loc[numerical_features].columns, cmap='coolwarm', center=0.20)

#### Il est intéressant de voir que budget et facenumber_in_poster sont très peu corrélées aux autres variables

# Cercle des corrélations et première réduction de dimensionalité des variables numériques

In [ ]:
import matplotlib.pyplot as plt
from matplotlib.collections import LineCollection
import numpy as np
import pandas as pd
from scipy.cluster.hierarchy import dendrogram

def display_circles(pcs, n_comp, pca, axis_ranks, labels=None, label_rotation=0, lims=None):
    for d1, d2 in axis_ranks: # On affiche les 3 premiers plans factoriels, donc les 6 premières composantes
        if d2 < n_comp:

            # initialisation de la figure
            fig, ax = plt.subplots(figsize=(30,30))

            # détermination des limites du graphique
            if lims is not None :
                xmin, xmax, ymin, ymax = lims
            elif pcs.shape[1] < 30 :
                xmin, xmax, ymin, ymax = -1, 1, -1, 1
            else :
                xmin, xmax, ymin, ymax = min(pcs[d1,:]), max(pcs[d1,:]), min(pcs[d2,:]), max(pcs[d2,:])

            # affichage des flèches
            # s'il y a plus de 30 flèches, on n'affiche pas le triangle à leur extrémité
            if pcs.shape[1] < 30 :
                plt.quiver(np.zeros(pcs.shape[1]), np.zeros(pcs.shape[1]),
                   pcs[d1,:], pcs[d2,:], 
                   angles='xy', scale_units='xy', scale=1, color="grey")
                # (voir la doc : https://matplotlib.org/api/_as_gen/matplotlib.pyplot.quiver.html)
            else:
                lines = [[[0,0],[x,y]] for x,y in pcs[[d1,d2]].T]
                ax.add_collection(LineCollection(lines, axes=ax, alpha=.1, color='black'))
            
            # affichage des noms des variables  
            if labels is not None:  
                for i,(x, y) in enumerate(pcs[[d1,d2]].T):
                    if x >= xmin and x <= xmax and y >= ymin and y <= ymax :
                        plt.text(x, y, labels[i], fontsize='14', ha='center', va='center', rotation=label_rotation, color="blue", alpha=0.5)
            
            # affichage du cercle
            circle = plt.Circle((0,0), 1, facecolor='none', edgecolor='b')
            plt.gca().add_artist(circle)

            # définition des limites du graphique
            plt.xlim(xmin, xmax)
            plt.ylim(ymin, ymax)

        
            # affichage des lignes horizontales et verticales
            plt.plot([-1, 1], [0, 0], color='grey', ls='--')
            plt.plot([0, 0], [-1, 1], color='grey', ls='--')

            # nom des axes, avec le pourcentage d'inertie expliqué
            plt.xlabel('F{} ({}%)'.format(d1+1, round(100*pca.explained_variance_ratio_[d1],1)))
            plt.ylabel('F{} ({}%)'.format(d2+1, round(100*pca.explained_variance_ratio_[d2],1)))

            plt.title("Cercle des corrélations (F{} et F{})".format(d1+1, d2+1))
            plt.show(block=False)
        
def display_factorial_planes(X_projected, n_comp, pca, axis_ranks, labels=None, alpha=1, illustrative_var=None):
    for d1,d2 in axis_ranks:
        if d2 < n_comp:
 
            # initialisation de la figure       
            fig = plt.figure(figsize=(16,9))
        
            # affichage des points
            if illustrative_var is None:
                plt.scatter(X_projected[:, d1], X_projected[:, d2], alpha=alpha)
            else:
                illustrative_var = np.array(illustrative_var)
                for value in np.unique(illustrative_var):
                    selected = np.where(illustrative_var == value)
                    plt.scatter(X_projected[selected, d1], X_projected[selected, d2], alpha=alpha, label=value)
                plt.legend()

            # affichage des labels des points
            if labels is not None:
                for i,(x,y) in enumerate(X_projected[:,[d1,d2]]):
                    plt.text(x, y, labels[i],
                              fontsize='14', ha='center',va='center') 
                
            # détermination des limites du graphique
            boundary = np.max(np.abs(X_projected[:, [d1,d2]])) * 1.1
            '''
            plt.xlim([-boundary,boundary])
            plt.ylim([-boundary,boundary])
            '''
            
            plt.xlim([-5,10])
            plt.ylim([-5,10])


            # affichage des lignes horizontales et verticales
            plt.plot([-100, 100], [0, 0], color='grey', ls='--')
            plt.plot([0, 0], [-100, 100], color='grey', ls='--')

            # nom des axes, avec le pourcentage d'inertie expliqué
            plt.xlabel('F{} ({}%)'.format(d1+1, round(100*pca.explained_variance_ratio_[d1],1)))
            plt.ylabel('F{} ({}%)'.format(d2+1, round(100*pca.explained_variance_ratio_[d2],1)))

            plt.title("Projection des films (coloration : à définir) (sur F{} et F{})".format(d1+1, d2+1))
            plt.show(block=False)

def display_scree_plot(pca):
    scree = pca.explained_variance_ratio_*100
    plt.bar(np.arange(len(scree))+1, scree)
    plt.plot(np.arange(len(scree))+1, scree.cumsum(),c="red",marker='o')
    plt.xlabel("rang de l'axe d'inertie")
    plt.ylabel("pourcentage d'inertie")
    plt.title("Eboulis des valeurs propres")
    plt.show(block=False)

def plot_dendrogram(Z, names):
    plt.figure(figsize=(10,25))
    plt.title('Hierarchical Clustering Dendrogram')
    plt.xlabel('distance')
    dendrogram(
        Z,
        labels = names,
        orientation = "left",
    )
plt.show()

In [ ]:

from sklearn import decomposition
from sklearn import preprocessing

# Import `PCA` from `sklearn.decomposition`
from sklearn.decomposition import PCA

# Build the model
pca = PCA(n_components=2)

# choix du nombre de composantes à calculer
n_comp = 6

# import de l'échantillon
data = df

# selection des colonnes à prendre en compte dans l'ACP
data_pca = df [numerical_features]

# préparation des données pour l'ACP
#data_pca = data_pca.fillna(data_pca.mean()) # Il est fréquent de remplacer les valeurs inconnues par la moyenne de la variable
data_pca = data_pca.dropna()

X = data_pca.values
#names = data["idCours"] # ou data.index pour avoir les intitulés

#features = data.columns
features = data_pca.columns

# Centrage et Réduction
std_scale = preprocessing.StandardScaler().fit(X)
X_scaled = std_scale.transform(X)

# Calcul des composantes principales
pca = decomposition.PCA(n_components=n_comp)
pca.fit(X_scaled)

# Eboulis des valeurs propres
display_scree_plot(pca)

# Cercle des corrélations
pcs = pca.components_
#plt.figure(figsize=(16,10))
plt.rcParams["figure.figsize"] = [16,9]
display_circles(pcs, n_comp, pca, [(0,1),(2,3),(4,5)], labels = np.array(features))


# Projection des individus
X_projected = pca.transform(X_scaled)
#display_factorial_planes(X_projected, n_comp, pca, [(0,1),(2,3),(4,5)], illustrative_var=data_pca[['nutrition_scoring']].values[:,0])
display_factorial_planes(X_projected, n_comp, pca, [(0,1),(2,3),(4,5)])


plt.show()



# Sélection des features pour le clustering

2 features à supprimer :
movie_imdb_link              1.000000
aspect_ratio                 0.934574 => information technique

facenumber_in_poster         0.997399  => voir la relation de ces variables avec les scores
num_user_for_reviews         0.995798  => voir la relation de ces variables avec les scores
num_critic_for_reviews       0.990196  => voir la relation de ces variables avec les scores





# Features à transformer / ajouter

Pour les variables qualitatives : créer autant de variables que de valeurs distinctes (one hot encode) ?

movie_title                  1.000000   => La distance entre chaque valeur devra être une distance de chaîne de caractère.   Mais comment faire un algo de clustering qui ne calcule pas la distance de la même façon pour cet attribut là que pour les autres ?  Faire un vecteur one hot avec le nombre distinct de titres de films dedans, et réduire sa dimensionalité  ?

# Features catégorielles

In [ ]:
qgrid_show(df[categorical_features_totransform])

In [ ]:
df['country'].unique().shape[0]

In [ ]:
print(f'{df.shape[0]} valeurs au total dans le dataframe')
for col in df[categorical_features_totransform]:
    print(f {col} : {df[col].unique().shape[0]} valeurs uniques')

# Annexe : inutile, à effacer plus tard

In [21]:
'''
df_transformed = df.copy(deep=True)

for feature_totransform in categorical_features_tosplit_andtransform:
    all_features = []

    for i, row in df.iterrows():
        if (type(row[feature_totransform]) == str):        
            features_list = row[feature_totransform].split(sep='|')
            for feature_name in features_list:
                all_features.append(feature_totransform+'_'+feature_name)

    all_features = list(set(all_features))

    for feature_name in all_features:
        df_transformed[feature_name] = 0


    for i, row in df.iterrows():
        if (type(row[feature_totransform]) == str):        
            features_list = row[feature_totransform].split(sep='|')
            for feature_name in features_list:
                df_transformed.at[i, feature_totransform+'_'+feature_name] = 1
'''

In [22]:
'''
df_transformed = df.copy(deep=True)

for feature_totransform in categorical_features_tosplit_andtransform:
    for i, row in df.iterrows():
        if (type(row[feature_totransform]) == str):        
            features_list = row[feature_totransform].split(sep='|')
            for feature_name in features_list:
                df_transformed.at[i, feature_totransform+'_'+feature_name] = 1
'''

KeyboardInterrupt: 